### PR TITLE
Simulation time speedup

### PR DESCRIPTION
**Pull request by RobinGeens**
There's a lot of low-hanging fruit in the optimization of the code.

Most of the simulation time is spent on dict-related operations, `__hash__` and `__eq__` in particular. This PR includes the following optimalizations:

- For layer dimensions: compute the hash at instantiation and replace the dynamic hash function with a reference to the static hash
- For layer dimensions: simplify `__eq__` to comparing hashes
- Replace `dict` with manually added default values for expected lookups with `defaultdict`

### CODE REVIEW

File: zigzag/datatypes.py [hunk 1]
```diff
@@ -9,13 +9,20 @@ class OperandABC(metaclass=ABCMeta):
     """! Abstract Base Class for all dimension- and operand-like classes"""
 
     def __init__(self, name: str):
-        self.name = name
+        self.__name = name
+        self.__hash = hash(name) ^ hash(type(self))
+
+    @property
+    def name(self):
+        """Protect the class variable from reassignment (as this would invalidate the stored hash value)"""
```
Comment (Andful) [inline zigzag/datatypes.py:19]: Frozen dataclasses might be interesting for this.

https://docs.python.org/3/library/dataclasses.html#frozen-instances

Comment (RobinGeens) [reply zigzag/datatypes.py:19]: On first sight, this doesn't seem to increase hash speed. Am I missing something?

**Comment by RobinGeens** [comment]
output

For reference: this is the relative computation time of all functions when running `example.py` on main. I used this command:
`python -m cProfile -o output.pstats example.py`
`gprof2dot --colour-nodes-by-selftime -f pstats output.pstats |  dot -Tpng -o output.png`

**Comment by RobinGeens** [comment]
Note: by optimizing the hash of `OperandABC`, the execution time of `example.py` decreased from 180s to 133s (1.35x speedup)

**Review by asyms — APPROVED** [review]

**Pull request merged**